numba/dppl/dparray.py

```diff
@@ -0,0 +1,660 @@
+#from ._ndarray_utils import _transmogrify
+import numpy as np
+from inspect import getmembers, isfunction, isclass, isbuiltin
+from numbers import Number
+import numba
+from types import FunctionType as ftype, BuiltinFunctionType as bftype
+from numba import types
+from numba.extending import typeof_impl, register_model, type_callable, lower_builtin
+from numba.np import numpy_support
+from numba.core.pythonapi import box, allocator
+from llvmlite import ir
+import llvmlite.llvmpy.core as lc
+import llvmlite.binding as llb
+from numba.core import types, cgutils
+import builtins
+import sys
+from ctypes.util import find_library
+from numba.core.typing.templates import builtin_registry as templates_registry
+from numba.core.typing.npydecl import registry as typing_registry
+from numba.core.imputils import builtin_registry as lower_registry
+import importlib
+import functools
+import inspect
+from numba.core.typing.templates import CallableTemplate
+from numba.np.arrayobj import _array_copy
+
+debug = False
+
+def dprint(*args):
+    if debug:
+        print(*args)
+        sys.stdout.flush()
+
+flib = find_library('mkl_intel_ilp64')
+dprint("flib:", flib)
+llb.load_library_permanently(flib)
+
+sycl_mem_lib = find_library('DPPLSyclInterface')
+dprint("sycl_mem_lib:", sycl_mem_lib)
+llb.load_library_permanently(sycl_mem_lib)
+
+import dpctl
+from dpctl._memory import MemoryUSMShared
+import numba.dppl._dppl_rt
+
+functions_list = [o[0] for o in getmembers(np) if isfunction(o[1]) or isbuiltin(o[1])]
+class_list = [o for o in getmembers(np) if isclass(o[1])]
+# Register the helper function in dppl_rt so that we can insert calls to them via llvmlite.
+for py_name, c_address in numba.dppl._dppl_rt.c_helpers.items():
+    llb.add_symbol(py_name, c_address)
+
+array_interface_property = "__array_interface__"
+def has_array_interface(x):
+    return hasattr(x, array_interface_property)
+
+class ndarray(np.ndarray):
+    """
+    numpy.ndarray subclass whose underlying memory buffer is allocated
+    with a foreign allocator.
+    """
+    def __new__(subtype, shape,
+                dtype=float, buffer=None, offset=0,
+                strides=None, order=None):
+        # Create a new array.
+        if buffer is None:
+            dprint("dparray::ndarray __new__ buffer None")
+            nelems = np.prod(shape)
+            dt = np.dtype(dtype)
+            isz = dt.itemsize
+            buf = MemoryUSMShared(nbytes=isz*max(1,nelems))
+            new_obj = np.ndarray.__new__(
+                subtype, shape, dtype=dt,
+                buffer=buf, offset=0,
+                strides=strides, order=order)
+            if hasattr(new_obj, array_interface_property):
```

Comment:
@diptorupd I thought that at some version of dpnp that I was looking at that it was named "__array_interface__".  This line 75 will only really be operative once we get to the dpnp merger phase.  We might want to do it as a __getattr__ so that we can dynamically form the object from shape, dtype, strides, etc that are stored in the regular ndarray class.  In short, this code is not finalized but is a placeholder.